### PR TITLE
Fixes IE11 issue with select losing focus having selected an item.

### PR DESCRIPTION
**Pull request by ToreOlavKristiansen**
This pull request includes a

- [ x] Bug fix
- [ ] New feature
- [ ] Translation

The following changes were made

- Changed from calling focus on the search input element to the select element.

If this is related to an existing ticket, include a link to it as  #4570 #3332.

### CODE REVIEW

**Comment by ToreOlavKristiansen** [comment]
I closed the pull request because one of the automated tests failed.

**Comment by ToreOlavKristiansen** [comment]
I did more testing and I don't see any issue even if the test fails. The input field might not have focus, but the select has had its focus call and thus everything works.

File: src/js/select2/selection/search.js [hunk 1]
```diff
@@ -175,7 +175,7 @@ define([
 
     this.resizeSearch();
     if (searchHadFocus) {
-      this.$search.focus();
+      this.$element.focus();
```
Comment (jeznag) [inline src/js/select2/selection/search.js]: Will this have side effects on other types of inputs? I did another PR which only modifies the behaviour for tag input as a precaution. https://github.com/select2/select2/pull/4958/files

Comment (ToreOlavKristiansen) [inline src/js/select2/selection/search.js]: $element.focus() invokes the focus function of the select2 plugin, not the "tag". This function should handle all cases so I don't think it has any side effects. We are using select elements in our solutions and the bug needs fixing for that too.

Did I understand you correctly?

Comment (alexweissman) [reply src/js/select2/selection/search.js]: It seems that there are several competing solutions to this problem:

https://github.com/select2/select2/issues/4398#issuecomment-281721604
https://github.com/select2/select2/pull/4958/files
https://github.com/select2/select2/commit/294e7e664d95eafb7f41264c2b87412ad9aa48fc

**Comment by alexweissman** [comment]
@ToreOlavKristiansen @jeznag can either of you confirm that this is still an issue in Select2 v4.0.5?  We fixed a more glaring issue with `focus` in 4.0.4., so I'm not sure if this is just another manifestation of that problem, or a separate issue specific to IE.  Thanks!

**Comment by ToreOlavKristiansen** [comment]
@alexweissman it is fixed in v4.0.5

**Comment by ToreOlavKristiansen** [comment]
@alexweissman it is improved in v4.0.5, but it is still a problem for multi-select: focus is lost when selecting an item in IE11. It works fine in Chrome.

**Comment by alexweissman** [comment]
Alright, I'll merge this in!

**Comment by alexweissman** [comment]
Actually, we need to fix the broken build.  If you believe that the failing tests do not reflect accurately on the expected behavior, then we need to change either your code or the tests 😛 

I notice that #4958 is handling the focus correctly, so maybe for your solution to work, we just need to add some logic for `searchHadFocus`?  We can still leave out their check for `isTagInput` if you believe that is not actually necessary.

**Comment by ToreOlavKristiansen** [comment]
I will test some more tomorrow. In my test earlier today I did not alter base.js and it looked like Chrome was ok without it. If that is the case, I will add another pull request.

**Comment by ToreOlavKristiansen** [comment]
I tested. We need both changes. I modified to please the unit tests. I also fixed an obvious duplicate text issue in the Norwegian translation.

**Comment by alexweissman** [comment]
Awesome, thank you!  Not sure what happened with the Norwegian translation, I must have performed a bad merge conflict resolution 😕

**Comment by ToreOlavKristiansen** [comment]
We use the select2 control a lot and really appreciate all your hard work on it. I am glad to contribute. Thanks :)

**Comment by alexweissman** [comment]
Merged into `develop` for the 4.0.6-rc.1 pre-release.

**Comment by alexweissman** [comment]
Actually, one more question - did you try just checking `searchHadFocus` instead of `isTagInput`?

**Comment by ToreOlavKristiansen** [comment]
I did. Then everything appears to be working fine, but the unit tests fail on a check for "search" having focus.

**Comment by alexweissman** [comment]
Ah ok, I understand.  Everything is wrapped in `if (searchHadFocus)` anyway 👍

**Comment by alexweissman** [comment]
@ToreOlavKristiansen I'm still seeing a possibly related problem in #3168 - can you take a look?

**Comment by ToreOlavKristiansen** [comment]
@alexweissman I do get this issue in IE11. I wouldn't say it is related to the lost focus issue though.

I see that the selected items are cleared first and then IE gets the backspace key event and goes back. Should we check for a backspace key event when clearing the items and prevent it from bubbling up?

**Comment by alexweissman** [comment]
We could try that, but I'd like to know why this only seems to affect IE 😕

**Comment by ToreOlavKristiansen** [comment]
I see that having used backspace, it looks like the control has focus, but is not possible to start typing in Chrome. So the search textbox does not have focus. IE then probably figures since no input field has focus and goes back when it gets the backspace. The issue could be fixed by calling this.$element.focus(); having cleared the items. Then it is perhaps possible to start typing too.

**Comment by alexweissman** [comment]
It would seem that there are other issues related to IE losing focus too easily (#2503, for example).  Maybe it has ADHD 🙂 

At any rate, #3394 claims to fix this.  Can you check it out?

**Comment by ToreOlavKristiansen** [comment]
No, that does not fix it.

I was able to fix it by adding this code as shown in the image:
        ```
evt.preventDefault();
var self = this;
window.setTimeout(function () {
    self.$element.focus();
}, 0);

```
image

**Comment by ToreOlavKristiansen** [comment]
The prevent default on the event fixes the IE browser back issue. The focus fixes not being able to start typing for both IE and Chrome.

**Comment by alexweissman** [comment]
Interesting...I wonder what #3394 is trying to achieve, then.  Seems like it is doing something similar, but in `search.js` instead of `allowClear.js`.

**Comment by ToreOlavKristiansen** [comment]
Let me know if you want a pull request for this.

**Comment by alexweissman** [comment]
Yes, that would be great!

**Comment by ToreOlavKristiansen** [comment]
Here it is: #5161

I made the mistake of including the dist files. Will you be able to sort that out?
I don't use git often. I see that all my changes are included in the pull request, not just the allowClear.js modifications. Please let me know if I am doing something wrong here.

**Comment by alexweissman** [comment]
Thanks!  Yes, it would be more helpful if the PR didn't include changes to `dist`.  However, if they are in a separate commit, I can just `cherry-pick` the src commits.

**Comment by nehachaudhary** [comment]
@ToreOlavKristiansen The fix you gave in 5161 doesn't work either.

**Pull request closed without merging**